docs/proposals/shuffle-sharding-on-the-read-path.md

```diff
@@ -0,0 +1,350 @@
+
+#### Edge case: queriers notice the ring topology change before distributors
+
+Consider the following scenario:
+
+1. Tenant A shard is composed by ingesters 1,2,3,4,5,6
+2. Tenant A is remote writing 1 single series and gets replicated to ingester 1,2,3
+3. The ring topology changes and tenant A shard is ingesters 1,2,3,7,8,9
+4. Querier notices the ring topology change and updates the in-memory map. Given tenant A series were only on ingester 1,2,3, the querier maps tenant A to ingester 1,2,3 (because of what received from ingesters via gRPC) and 7,8,9 (because of the current state of the ring)
+5. Distributor hasn’t updated the ring state yet
+6. Tenant A remote writes 1 **new** series, which get replicated to 4,5,6
+7. Distributor updates the ring state
+8. **Race condition:** querier will not know that ingesters 4,5,6 contains tenant A data until the next sync
+
+### Option #2: use only the information contained in the ring.
+
+*This section describes an alternative approach.  Discussion is still on-going.*
+
+The idea is for the queries to be able to deduce what ingesters could possibly hold data for a given tenant by just consulting the ring (and the per-tenant sub ring sizes).  We posit that this is possible with only a single piece of extra information: a single timestamp per ingester saying when the ingester first joined the ring.
+
+#### Scenario: ingester scale up
+
+When a new ingester is added to the ring, there will be a set of user subrings that see a change: an ingester being removed, and a new one being added.  We need to guarantee that for some time period (the block flush interval), the ingester removed is also consulted for queries.
+
+To do this, during the subring selection if we encounters an ingester added within the time period, we will add this to the subring but continue node selection as before - in effect, selecting an extra ingester:
+
+```go
+var (
+    subringSize   int
+    selectedNodes []Node
+    deadline      = time.Now().Add(-flushWindow)
+)
+
+for len(selectedNodes) < subringSize {
+    token := random.Next()
+    node := getNodeByToken(token)
+    for {
+        if node in selectedNodes {
+            node = node.Next()
+            continue
+        }
+        if node.Added.After(deadline) {
+            subringSize++
+            selectedNodes.Add(node)
+            node = node.Next()
+            continue
+        )
+        selectedNodes.Add(node)
+        break
+    )
+}
+```
+
+#### Scenario: ingester scale down
+
+When an ingester is permanently removed from the ring it will flush its data to the object store and the subrings containing the removed ingester will gain a “new” ingester.  Queries consult the store and merge the results with those from the ingesters, so no data will be missed.
+
+TODO do we need “leaving” ingester to hang around for 15mins after flushing until the store gateway is guaranteed to pick up their flush blocks?
```

Comment:
I would change this `TODO` with the actual statement in the original google doc comments.

Reply:
I've removed TODO and put this text in:

> Queriers and store-gateways will discover newly flushed blocks on next sync (`-blocks-storage.bucket-store.sync-interval`, default 5 minutes). Multiple ingesters should not be scaled-down within this interval.
>
> To improve read-performance, queriers and rulers are usually configured with non-zero value of `-querier.query-store-after` option. This option makes queriers and rulers to only consult ingesters when running queries withing specified time window (eg. 12h). During scale-down this needs to be lowered in order to let queriers and rulers use flushed blocks from the storage.

Reply:
LGTM, thanks!